src/expr/symbol_table.cpp

```diff
@@ -255,18 +314,121 @@ void SymbolTable::Implementation::reset() {
 }
 
+void SymbolTable::Implementation::bindWithOverloading(const string& name, Expr obj) {
+  CDHashMap<string, Expr>::const_iterator it = d_exprMap->find(name);
+  if(it != d_exprMap->end()) {
+    if((*it).second!=obj) {
+      if(!isOverloadedFunction((*it).second)) {
+        // mark previous as overloaded
+        markOverloaded(name, (*it).second);
+      }
+      // mark this as overloaded
+      markOverloaded(name, obj);
+    }
+  }
+}
+
+void SymbolTable::Implementation::markOverloaded(const string& name, Expr obj) {
+  Trace("parser-overloading") << "Overloaded function : " << name;
+  Trace("parser-overloading") << " with type " << obj.getType() << std::endl;
+  d_overloaded_symbols->insert(obj);
+  // get the argument types
+  Type t = obj.getType();
+  Type rangeType = t;
+  std::vector< Type > argTypes;
+  if(t.isFunction()) {
+    argTypes = ((FunctionType)t).getArgTypes();
+    rangeType = ((FunctionType)t).getRangeType();
+  }else if(t.isConstructor()) {
+    argTypes = ((ConstructorType)t).getArgTypes();
+    rangeType = ((ConstructorType)t).getRangeType();
+  }else if(t.isTester()) {
+    argTypes.push_back( ((TesterType)t).getDomain() );
+    rangeType = ((TesterType)t).getRangeType();
+  }else if(t.isSelector()) {
+    argTypes.push_back( ((SelectorType)t).getDomain() );
+    rangeType = ((SelectorType)t).getRangeType();
+  }
+  // add to the trie
+  TypeArgTrie * tat = &d_overload_type_arg_trie[name];
+  for(unsigned i=0; i<argTypes.size(); i++) {
+    tat = &(tat->d_children[argTypes[i]]);
+  }
+  tat->d_symbols[rangeType] = obj;
+}
+  
+bool SymbolTable::Implementation::isOverloadedFunction(Expr fun) const {
+  return d_overloaded_symbols->find(fun)!=d_overloaded_symbols->end();
+}
+
+Expr SymbolTable::Implementation::getOverloadedConstantForType(const std::string& name, Type t) const {
+  std::unordered_map< std::string, TypeArgTrie >::const_iterator it = d_overload_type_arg_trie.find(name);
+  if(it!=d_overload_type_arg_trie.end()) {
+    std::map< Type, Expr >::const_iterator its = it->second.d_symbols.find(t);
+    if(its!=it->second.d_symbols.end()) {
+      return its->second;
+    }
+  }
+  return d_nullExpr;
+}
+
+Expr SymbolTable::Implementation::getOverloadedFunctionForTypes(const std::string& name, 
+                                                                const std::vector< Type >& argTypes) const {
+  std::unordered_map< std::string, TypeArgTrie >::const_iterator it = d_overload_type_arg_trie.find(name);
+  if(it!=d_overload_type_arg_trie.end()) {
+    const TypeArgTrie * tat = &it->second;
+    for(unsigned i=0; i<argTypes.size(); i++) {
+      std::map< Type, TypeArgTrie >::const_iterator itc = tat->d_children.find(argTypes[i]);
+      if(itc!=tat->d_children.end()) {
+        tat = &itc->second;
+      }else{
+        // no functions match 
+        return d_nullExpr;
+      }
+    }
+    // now, we must ensure that there is one and only one active symbol at this node
```

Comment:
How can it happen that there is more than one active symbol at a node?  Shouldn't that be an error?

Reply:
Technically it doesn't have to be.  To demonstrate the design decision here, there is no argument that this should be an error:

(declare-fun f (Int) Bool)
(declare-fun f (Int) Int)
(assert (= (f 0) 4)

since we are *using* f in a way that requires type inference.

However, if we just declare the functions:
(declare-fun f (Int) Bool)
(declare-fun f (Int) Int)
...does not refer to f...

then this will not throw an error currently.

The decision is whether to throw an error at declaration time or at use time.  Potentially once the higher-order branch is merged we will have the ability to use declarations above using function type casts:
(declare-fun f (Int) Bool)
(declare-fun f (Int) Int)
(assert (= ((as f (-> Int Int)) 0) 4)

hence we want to allow potentially illegal declarations but as long as the user does things right we don't complain.